include/deal.II/physics/notation.h

```diff
@@ -0,0 +1,1550 @@
+                                   const bool         symmetric)
+        {
+          if (symmetric == true)
+            {
+              Assert((component_n < SymmetricTensor<2,2>::n_independent_components),
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,2>::n_independent_components));
+            }
+          else
+            {
+              Assert((component_n < Tensor<2,2>::n_independent_components),
+                     ExcIndexRange(component_n,0,Tensor<2,2>::n_independent_components));
+            }
+
+          switch (component_n)
+            {
+            case (0):
+              return std::make_pair(0u,0u);
+            case (1):
+              return std::make_pair(1u,1u);
+            case (2):
+              return std::make_pair(0u,1u);
+            case (3):
+              Assert(symmetric == false,
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,2>::n_independent_components));
+              return std::make_pair(1u,0u);
+            default:
+              break;
+            }
+
+          AssertThrow(false, ExcInternalError());
+          return std::pair<unsigned int, unsigned int> ();
+        }
+
+
+        template<>
+        std::pair<unsigned int, unsigned int>
+        indices_from_component<3> (const unsigned int component_n,
+                                   const bool         symmetric)
+        {
+          if (symmetric == true)
+            {
+              Assert((component_n < SymmetricTensor<2,3>::n_independent_components),
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,3>::n_independent_components));
+            }
+          else
+            {
+              Assert((component_n < Tensor<2,3>::n_independent_components),
+                     ExcIndexRange(component_n,0,Tensor<2,3>::n_independent_components));
+            }
+
+          switch (component_n)
+            {
+            case (0):
+              return std::make_pair(0u,0u);
+            case (1):
+              return std::make_pair(1u,1u);
+            case (2):
+              return std::make_pair(2u,2u);
+            case (3):
+              return std::make_pair(1u,2u);
+            case (4):
+              return std::make_pair(0u,2u);
+            case (5):
+              return std::make_pair(0u,1u);
+            case (6):
+              Assert(symmetric == false,
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,3>::n_independent_components));
+              return std::make_pair(1u,0u);
+            case (7):
+              Assert(symmetric == false,
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,3>::n_independent_components));
+              return std::make_pair(2u,0u);
+            case (8):
+              Assert(symmetric == false,
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,3>::n_independent_components));
+              return std::make_pair(2u,1u);
```

Comment:
Assuming these exceptions here are duplicative, then you end up with a case where you could as well have done everything table-based, i.e., using
```
  static const unsigned int indices[9][2] = {{0,0}. {1,1}. ...};
  return std::make_pair(indices[component_n][0], indices[component_n][1]);
```
That's easier to read, and I bet it's also easier for the compiler to optimize. I suspect that you could do something similar for the other functions above/below this one.

Reply:
I do think its funny that the one and only time I choose to use a `switch` statement (which I try to avoid out of a dislike of the syntax) it ends up being removed. Ironic :-)

Reply:
:-) I do think that if something can be expressed as a table, then that's worth it. It's another thing that makes it so much simpler for a compiler to optimize stuff, and this seems like a place where you'd like the compiler to optimize things well, given that it happens in the hot path of assembly.